### ISSUE TITLE
Question: Using LiquidHaskell as a library for checking a custom language

### ISSUE THREAD

**Issue by chrisdone** [ucsd-progsys/liquidhaskell#1768]
Hi, just a quick question.

I have a typed language which is a trivial, strict subset of Haskell/ML which has basic arithmetic, lambdas, letrec (so, mutual recursion) and type-classes with an `Integer` type and a `Fixed n` type (no fixed-width number types), arrays (arbitrary-sized). It has row types, but I could probably eliminate them for the purpose of calling LH. It's possible to run on the AST with type-class constraints, or the core-like version with dictionary passing.

I would be interested in checking:

1. Top priority: Impossibility of division by zero.
2. Low priority: Termination/totality.
3. Lowest priority: Possibly, in predicting ahead of time how large these arbitrary-precision numbers could grow. But just if that is available.

Users will not be specifying their own properties. I would build LH into the language such that the arithmetic operation `/` has the proper constraint, and a `if x/=0 ..` or `case n of 0 -> ..; n -> ...` would provide the proof. Use of LiquidHaskell would be like a linter to say "hey, I'm not going to block you from running your code, but I'm warning you about this situation." 

What's your estimation of difficulty in using LiquidHaskell on my AST? @ranjitjhala @nikivazou 

Cheers!

**Comment by ranjitjhala** [comment]
Hi @chrisdone, this sounds like a fun project! 

I am unclear about one crucial thing -- are you using e.g. the GHC API, 
and literally the same datatypes `CoreExpr` etc. or do you have your 
own standalone infrastructure? If its the _same_ core then its likely 
very easy to use LH. If not, I think you'd have to translate your stuff 
into GHC Core to use the rest of LH.

At a high-level LH has three parts:

1. "Front-end" which parses  in  user-written *specifications*
2. "Middle-end" (?) which combines (1) with `core` to generate *constraints*
3. "Back-end" which *solves* the constraints.

So if you already have / can generate GHC core, it looks like you could reuse (2), (3) as is.

If you don't, i.e. if you have your own custom AST then we'd have to do the constraint generation.
In this case, you may be interested in 

https://github.com/ranjitjhala/sprite-lang

which is a tiny implementation of LH style refinement types for a core ML-like language of the 
sort you're describing (e.g. here's the https://github.com/ranjitjhala/sprite-lang/blob/master/src/Language/Sprite/L5/Types.hs#L143-L153) -- its a small self-contained repo that has all three steps above, so 
may be a good starting point. There's also an accompanying document that explains each of 
languages and how the constraint generation works. 

So to sum up: if you're using / can translate to GHC Core then super easy, but if not we'll have 
to redo constraint generation for your AST (as explained in the "sprite" repo) -- the actual *solving*
is factored out into a separate package `liquid-fixpoint` that both LH and `sprite` use.

Does that make sense?

**Comment by chrisdone** [comment]
Wow, what a helpful reply! :clap::smiley: 

> I am unclear about one crucial thing -- are you using e.g. the GHC API,
> and literally the same datatypes CoreExpr etc. or do you have your
> own standalone infrastructure? If its the same core then its likely
> very easy to use LH. If not, I think you'd have to translate your stuff
> into GHC Core to use the rest of LH.

Right, I have my own AST. In that case, if your advise is to
translate, that's workable. GHC is a large and changing dependency
with a heavy foot-print to it. But it's not a show-stopper for our
project. I've worked with GHC's STG module before, so there's that.

> At a high-level LH has three parts:
> 
> "Front-end" which parses in user-written specifications
> "Middle-end" (?) which combines (1) with core to generate constraints
> "Back-end" which solves the constraints.
>
> So if you already have / can generate GHC core, it looks like you
> could reuse (2), (3) as is.

Reasonble pipeline. That matches my hope of doing (2) and (3).

> If you don't, i.e. if you have your own custom AST then we'd have to
> do the constraint generation.
>
> In this case, you may be interested in
> 
> https://github.com/ranjitjhala/sprite-lang
> 
> which is a tiny implementation of LH style refinement types for a core ML-like language of the
> sort you're describing (e.g. here's the https://github.com/ranjitjhala/sprite-lang/blob/master/src/Language/Sprite/L5/Types.hs#L143-L153) -- its a small self-contained repo that has all three steps above, so
> may be a good starting point. 

Great coincidence that you posted this recently, that'll be a very
handy reference! Your AST is very similar to mine modulo the row types.
 
> So to sum up: if you're using / can translate to GHC Core then super
> easy, but if not we'll have to redo constraint generation for your
> AST (as explained in the "sprite" repo) -- the actual solving is
> factored out into a separate package liquid-fixpoint that both LH
> and sprite use.
 >
> Does that make sense?

It does! Looks like I can:

1. Do a prototype of the idea by translating my language AST to Core
   and feeding it into LH, and treat LH more-or-less as a black box.
2. If that shows promise as a good fit with my language, I can look at
   a more tightly-coupled integration via liquid-fixpoint at the price
   of needing to know more about the workings of LH.

Thanks very much!

I'll let you know how it goes when I can allocate time to try it.
